### ISSUE TITLE
Plans for bitemporal queries and applications?

### ISSUE THREAD

**Issue by dangdennis** [tigerbeetle/tigerbeetle#157]
Is tigerbeetle planning to support bitemporal queries and execution such that we can retroactively apply debits/credits? Likewise, will we be able to track when such a retroactive action took place?

**Comment by sentientwaffle** [comment]
This isn't on the road map right now. But I'm interested to hear more about the use cases and applications for this feature.

**Comment by dangdennis** [comment]
Hi @sentientwaffle. Maybe this idea is entirely against the rules of double entry accounting. Happy to be educated!

The feature in discussion is essentially built-in tracking of all historical changes, including edits to invalid transactions or perhaps adding a transaction into the past. It's what [xtdb](https://docs.xtdb.com/concepts/bitemporality/) offers but they don't have the double entry accounting built in, nor your other features. In my head, bitemporarily = ability to query against "valid times" and "transaction time" (typically system time as of transaction creation). xtdb also follows an document model instead of a table model. Hopefully I can translate a decent but very common example.

Primary use-case is when a mistake arises in a transaction: perhaps incorrect amount, incorrect time. In all cases, we can cleanly revert it by appending a new transaction. But it's not as neat as being able to delete or edit the incorrect transaction entirely. In this mutation though, we should naturally have a log of the action for auditability.

**Example with an append-only logic**
Day 1 - Balance $20.
Day 2 - incorrect debit transaction occurs. Balance is now $10. Should be $20. 
Day 3 - A new credit transaction is applied to revert the debit. Balance is now $20.

This results in an annoying history. At least we can probably relate the pair of reverts and hide from clients. Is this what you'd normally do in the tigerbeetle db, and for normal double entry accounting?

**Example with tracked mutation possible**
Day 1 - Credit an account with $20. Balance $20.
Day 2 - incorrect transaction occurs. Balance is now $10. Should be $20. 
Day 3 - Delete the incorrect transaction. Balance is now $20. The deletion action is logged in the data model itself.

We can query for the account's balance with valid time of Day 3 for a balance of $20. If we also query for the account's transactions with the same parameters, we don't see the incorrect transaction nor the revert transaction. 

For an easy audit of the account history, we'd like to know why and how an account went from $20 to $10 to $20 by Day 3. We can query for transactions with transaction time by Day 3 to see all transactions including the invalid transaction made on Day 2.

**Comment by sentientwaffle** [comment]
Thank you for the explanation!

> This isn't on the road map right now.

I just talked to @jorangreef and I was wrong — we may support this in the future after all.

Your "Example with an append-only logic" accurately reflects how it works right now in TigerBeetle — if there's a mistake, just create a new transfer to fix the mistake.

It sounds like bitemporal queries can be made to work with double-entry accounting. But I haven't wrapped my head around them so I can't really elaborate yet :grin:.

**Comment by rrffkk** [comment]
> It sounds like bitemporal queries can be made to work with double-entry accounting. But I haven't wrapped my head around them so I can't really elaborate yet 😁.
I've recently explored TigerBeetle, which made me reflect on the database landscape for financial transactions. It's interesting that for decades, organizations had to repeatedly implement debit/credit accounting patterns in general-purpose relational databases, often reinventing similar solutions with varying degrees of success. TigerBeetle's emergence as a specialized database for financial transactions seems to fill this historical gap. :D

While TigerBeetle offers interesting and well-thought-out concepts, there are some technical considerations that may affect its adoption in wider financial/accounting scenarios.

I think the absence of bitemporal logic becomes more significant given that all TigerBeetle records are immutable. While immutability makes sense for live financial transactions, its application in traditional financial accounting is more nuanced. In standard accounting practices, true immutability typically only applies after month-end closure; before that point, most financial accounting systems maintain entries in a "draft" state that can be modified until the month is closed.
Of course, both approaches require careful analysis.

BTW: Actually, there can even be a need for tritemporal logic. A great read on this is in Wikipedia: https://en.wikipedia.org/wiki/Decision_time

**Comment by jorangreef** [comment]
_TLDR: Bitemporality cuts not only across the recording layer but also the policy and reporting layers beyond TigerBeetle. The principle is to keep primitives in TigerBeetle and policy in your application. Correcting entries should be “added as information” with `user_data_64` to store the effective timestamp, run through explicit ring-fenced control accounts or ledgers, and then joined with two queries (one against the bulk of realtime transactions, the other against the smaller set of ring-fenced correcting entries) in reporting. Given that all queries in TigerBeetle are guaranteed to run with predictable complexity linear in the size of the result set, there is no material impact on performance with this recommended solution._

We've been thinking about this since '22, talking with companies to hear their experiences:

- banks doing batch processing
- fintechs moving from batch to realtime
- low volume users (QuickBooks)
- high volume users (on the order of 1 billion txns per day)
- traditional accountants in the real world
- engineers who've built well-known ledgers

It's been fascinating to get several perspectives.

Learnings so far:

Perhaps the biggest paradigm shift in thinking about bitemporality is to acknowledge that “there are multiple layers”.

In other words, it's not only TigerBeetle (recording layer). There are also at least the policy and reporting layers to consider.

Once one sees this, that bitemporality is bigger than TigerBeetle, that it's also about policy and reporting then it frees you up to tease the problem apart and decouple.

With bitemporality, therefore, it's not only a question of primitives but of policy, for example:

- By when must a product/business legally recognize adjustments? What's the limited duration of the window period during which changes are allowed, but thereafter not? This policy can differ according to asset class.

- Must these adjustments (or correcting entries) be separately recorded and explicitly ring-fenced?

- What limits and escalation procedures apply to error corrections? Who signs off on them?

- What's the knock on impact? What's the organization's policy around how correcting entries impact financial periods, whether past or future? For example, should correcting entries be allowed to increase a net debit cap limit on a clearing account or only interest earned?

And that's without diving into the reporting layer. For example, we've spoken with companies who routinely record up to 5 dates per transaction!

A second paradigm shift is to acknowledge that bitemporality and immutability are logically orthogonal concepts, not in tension when applied correctly.

Then, there are questions of whether to solve bitemporality on the write path or read path. The disadvantage with solving bitemporality entirely in the write path is that policy becomes coupled with recording. Whereas, if you record only the essential information in the write path(as/when learned), how you join this later in the read path in reporting can be flexible (there’s no data to rewrite). For example, you can decide to group bitemporal entries with the original corrected entry, or allow a user to click in and ungroup to show all entries associated with the logical business transaction.

Furthermore, there are questions of scale and where things are heading:

- TigerBeetle solves the primary problem, not of QuickBooks or ad hoc processing, but of high volume realtime transaction processing. If you can solve high resolution realtime, then you can always layer on things with lower resolution if you want to, but it's hard to go the other way.

- In the past, the world had to do batch processing, at least because OLTP performance couldn’t keep up. This implied additional reconciliation infrastructure. But things are moving to realtime transaction processing, where ad hoc entries are the 1% exception not the 99% norm.

Finally, to maximize the power-to-weight ratio of TigerBeetle, to minimize product surface area, to preserve simplicity and orthogonality and to avoid TigerBeetle becoming bloated, whenever we add a feature or primitive we first ask the question: Is this use case simply impossible at present? Or can it be achieved by composing primitives?

Taking these perspectives into account, therefore, when it comes to bitemporality, I think it's unlikely that we're going to see this as being solved entirely within TigerBeetle.

Our recommended solution is as follows:

- TigerBeetle is the recording layer.

- Bitemporality, for reasons of policy, is best solved in the reporting layer.

- Acknowledging the existence of both layers makes for a more flexible core long term, especially as a company evolves or grows new product lines, and as policies change across these.

- Correcting entries should be explicit and ring-fenced, and run through a separate control account or ledger.

- The window period, for when correcting entries may be applied to a previous financial period, should be explicitly bounded by policy.

- Assuming that 99% of transactions are correct (and not correcting entries), it's efficient at reporting time, to query the realtime accounts/ledger, and then join this with a time-bound query against the correcting accounts/ledger. In most cases, you can do this with only 2 queries against TigerBeetle, which is performant, even as you scale to billions of transactions—there's an explicit limit on everything here (by design in this solution).

- And since most transactions are chronological, that's the sort order we optimize for in TigerBeetle. We don't want to have to re-sort 99% of transactions—if only 1% are going to be moving around—since this would impact performance _dramatically_. Instead, given that 1% of correcting entries are time bound, and that the query to retrieve and sort these in reporting is also linear and limited, the sort is better done on the read path than on the write path.

- All this doesn't mean that bitemporality doesn't demand anything from TigerBeetle as recording layer. Since the issue was opened, and as @sentientwaffle alluded to above, we've added the ability for additional dates, apart from the recording date (`Transfer.timestamp`), to be recorded in `Transfer.user_data_64`, and multiple transfers to be linked together so you can record as many dates as you want, and group/sort on these in reporting.

The advantage of this solution is that there's nothing hidden or implicit. Everything is explicit, and limited for safety and performance.

While it's more work, the resulting system would be both more flexible and more correct—the application can vary specific bitemporality rules depending on exact circumstances, but, no matter what it does, there's always going to be a full immutable record with all the correcting transfers inside TigerBeetle.

**Comment by filmackay** [comment]
In bitemporal, the concept of a balance bifucates. You have the commit and effective view, and conceptually the real-world "balance" aligns more with the effective view. That is, on the path to the "eventual" balance - on that both views will agree.

This means:

- Balance is derived from effective (not commit) time. TB can't help here.
- Balance is a timeseries that can extend into the future. Avoiding negative balances now requires you to validate that the entire timeseries stays positive. Again, you don't want TB doing this!

I think you're right @jorangreef - it's a reporting problem, I don't think you want to contaminate/drag on TB.

If you now think of having commit-balance (kinda irrelevant?) and effective-balance; TB operates on commit-balance. Users, however, think of balances in terms of effective-balance - that's the whole point of correcting "history" (a reporting view). In that context I think the following concepts become less/non-useful:

- Balancing transfers
- `must_not_exceed` constraints

The moral equivalent of the above in effective balancing would involve them operating on entire time-series rather than aggregate values. You also would need to think about a transaction having its own time-series, needing to be overlaid on top of the ledger time-series. Utter madness from a performance/complexity perspective. Actually, this is a question: would you allow effective-at to vary by leg, or only at the transaction level? There are certainly use-cases for atomic transactions that have different effective-at's within them. They're kind of sub-transaction groups - that need to balance within themselves, but you want them applied atomically with the other groups that are operating on different effective-at. 

I wonder if there is a defensive approach that could be conducted on top of reporting. I guess that is not possible with un-coordinated transaction generators. In this world I feel balances become something that TB can't relate to, and that is a problem when you want balance protections to be enforced by TB.

All TB would be able to enforce is that the balance is eventually OK, it can't prevent defecit paths en-route.
